### PR TITLE
Use 141 toolchain for Windows build

### PR DESCRIPTION
**Pull request by matthew-brett**
See discussion at https://github.com/MacPython/numpy-wheels/issues/145

I'm afraid I can't test this because I don't have Azure Pipeline permissions
for my own account.

Do I need some permissions from y'all?

### CODE REVIEW

**Comment by charris** [comment]
> I'm afraid I can't test this because I don't have Azure Pipeline permissions
for my own account.

Test it where? It is running here.

**Comment by charris** [comment]
Linking is failing
```
MSVCRT.lib(loadcfg.obj) : error LNK2001: unresolved external symbol __enclave_config
_configtest.exe : fatal error LNK1120: 1 unresolved external
```

**Comment by matthew-brett** [comment]
Sorry - I meant - it is hard to test the code before doing a PR.

**Comment by matthew-brett** [comment]
It looks like the compilation step is failing to find the MS compilers...  I'm afraid this is going to mean a whole slew of cruddy debug commits :(

**Comment by charris** [comment]
For a bunch of cruddy test commits it helps to comment out all the other tests, especially travis.

**Comment by matthew-brett** [comment]
@isuruf - any ideas what might be going on here?    The error is the one Chuck posted above, repeated here:

```
MSVCRT.lib(loadcfg.obj) : error LNK2001: unresolved external symbol __enclave_config
_configtest.exe : fatal error LNK1120: 1 unresolved external
```

The configuration batch output is [here](https://dev.azure.com/numpy/27346c6a-2774-4eac-bf85-e068127c0ccc/_apis/build/builds/22439/logs/7) - it appears to be finding the 141 toolset and SDK correctly.

I've tried on `windows-latest` and `windows-2019`.

I can't replicate it locally.   I've forced the same SDK and am using the same toolset locally - and I've run out of things I can think of to try, as of about 1am in the morning.   Any ideas what I could try next?

**Comment by matthew-brett** [comment]
OK - I think I worked it out - was missing `set MSSdk=1`.  Ouch.

**Comment by charris** [comment]
The OSX failures started 10 days or so ago. I suspect azure is reponsible somehow, an upgrade from 10.9 to 10.14 may be responsible for the error message, I don't really understand that bit. The windows failures (and errors) are new and look like mostly f2py and may be mingw related.

**Comment by isuruf** [comment]
Windows failures are because the testing script is run with MinGW bash which adds it's compilers to the front of PATH and unfortunately there's a `link.exe` from MinGW toolchain and distutils tries to use that instead of the one from Visual Studio because `DISTUTILS_USE_SDK=1`

**Comment by matthew-brett** [comment]
What's the right thing to do?  We could unset `DISTUTILS_USE_SDK` before that step- but is that the right fix?

**Comment by charris** [comment]
The OSX failures occur after an azure update to the Python tools version from 0.193 to 0.197, apparently the OSX version changed in the process.  See "set python version" tab.

**Comment by isuruf** [comment]
> We could unset DISTUTILS_USE_SDK before that step- but is that the right fix?

Yeah, that works. It also tests that numpy built with v141 toolchain can successfully link with v142 toolchain.

**Comment by matthew-brett** [comment]
Simple-minded unset didn't seem to work : 5481094ca4be0afbe34e8df49d49e6bf76a5b3e4

**Comment by matthew-brett** [comment]
@isuruf - thanks - it seems that Bash is not picking up the unset, although it has clearly picked up the earlier variable set (any explanations welcome).   I've tried using the ugly Pipeline syntax to change the environment variables.

**Comment by charris** [comment]
I trying to figure out how to give you permissions. If you want to run on your own account, it may be something you can do for yourself.

**Comment by matthew-brett** [comment]
Most unfortunately, Disutils does this check for the relevant variables:

```
if "DISTUTILS_USE_SDK" in os.environ and "MSSdk" in os.environ and self.find_exe("cl.exe"):
```

https://github.com/pypa/setuptools/blob/main/setuptools/_distutils/msvccompiler.py#L250

This means that it doesn't matter what values these variables have, including empty strings or 0, it will still trigger effects in Distutils.

So, I have `unset` the variables in the Bash install / test step.  I think that's as neat as we can be, but I'm happy to hear alternatives!

**Comment by rgommers** [comment]
If it makes life easier (/possible), patching `MSVCCompiler.initialize` in https://github.com/numpy/numpy/blob/main/numpy/distutils/msvccompiler.py#L43 by copying over the relevant `else` code and taking out that side effect is fine with me.

**Comment by matthew-brett** [comment]
OK - I think this one is ready to go.  Windows tests passing, macOS failures unrelated.

**Comment by charris** [comment]
Thanks Matthew.

**Comment by matthew-brett** [comment]
Thanks Chuck.

Now - would y'all consider tiny little PRs to build 141 wheels with a build tag for 1.22.0 and 1.22.1 ?  That way, Scipy can assume that the builds are 141 compatible, at least for now.

**Comment by matthew-brett** [comment]
Oh - and thanks @isuruf, as ever, for your invaluable help.

File: azure/set_vs_toolchain.cmd [hunk 1]
```diff
@@ -0,0 +1,71 @@
+@@echo on
```
Comment (isuruf) [inline azure/set_vs_toolchain.cmd:1]: Can you add something like
```
REM Taken from conda-forge recipes licensed as follows
REM BSD-3-Clause
REM Copyright conda-forge contributors
```

**Comment by rgommers** [comment]
> Now - would y'all consider tiny little PRs to build 141 wheels with a build tag for 1.22.0 and 1.22.1 ? That way, Scipy can assume that the builds are 141 compatible, at least for now.

We should first make a `1.22.2` release with already merged PRs labelled for backport with this toolchain, there are more critical issues in there for SciPy. Maybe it makes sense to do `.post0` releases for `1.20.0`/`1.20.1`, I'm not sure - but it's not all that important either way once `1.22.2` has what we need.

File: azure/set_vs_toolchain.cmd (hunk 1 above)
Comment (matthew-brett) [reply azure/set_vs_toolchain.cmd:1]: Good point - see: https://github.com/MacPython/numpy-wheels/pull/150

**Comment by mattip** [comment]
Maybe we should port this to the numpy/numpy repo and use it in CI
- to make sure we are testing it
- which will make it easier to port to the cibuildwheel builds

Other projects may want to use this too

**Pull request merged**